include/engine/data_watchdog.hpp

```diff
@@ -0,0 +1,84 @@
+#include "engine/datafacade/shared_datafacade.hpp"
+
+#include "storage/shared_datatype.hpp"
+#include "storage/shared_memory.hpp"
+
+#include <boost/thread/shared_mutex.hpp>
+#include <boost/thread/lock_types.hpp>
+
+#include <memory>
+
+namespace osrm
+{
+namespace engine
+{
+
+// This class monitors the shared memory region that contains the pointers to
+// the data and layout regions that should be used. This region is updated
+// once a new dataset arrives.
+//
+// TODO: This also needs a shared memory reader lock with other clients and
+// possibly osrm-datastore since updating the CURRENT_REGIONS data is not atomic.
+// Currently we enfore this by waiting that all queries have finished before
+// osrm-datastore writes to this section.
+class DataWatchdog
+{
+  public:
+    DataWatchdog()
+        : shared_regions(storage::makeSharedMemory(
+              storage::CURRENT_REGIONS, sizeof(storage::SharedDataTimestamp), false, false)),
+        current_timestamp {storage::LAYOUT_NONE, storage::DATA_NONE, 0}
+    {
+    }
+
+    // Tries to connect to the shared memory containing the regions table
+    static bool TryConnect()
+    {
+        return storage::SharedMemory::RegionExists(storage::CURRENT_REGIONS);
+    }
+
+    // Check if it might be worth to try to aquire a exclusive lock
+    bool HasNewRegion() const
+    {
+        const boost::shared_lock<boost::shared_mutex> lock(current_timestamp_mutex);
+
+        const auto shared_timestamp =
+            static_cast<const storage::SharedDataTimestamp *>(shared_regions->Ptr());
```

Comment:
Hm, can 1/ the shared_regions unique_ptr ever be empty, 2/ the ptr returned here ever be a nullptr?

(same below)